vizro-ai/src/vizro_ai/dashboard/_response_models/page.py

```diff
@@ -0,0 +1,225 @@
+"""Page plan model."""
+
+import logging
+from collections import Counter
+from typing import List, Union
+
+try:
+    from pydantic.v1 import BaseModel, Field, PrivateAttr, ValidationError, root_validator, validator
+except ImportError:  # pragma: no cov
+    from pydantic import BaseModel, Field, PrivateAttr, ValidationError, root_validator, validator
+import vizro.models as vm
+from tqdm.auto import tqdm
+from vizro_ai.dashboard._response_models.components import ComponentPlan
+from vizro_ai.dashboard._response_models.controls import ControlPlan
+from vizro_ai.dashboard._response_models.layout import LayoutPlan
+from vizro_ai.dashboard.utils import _execute_step
+
+logger = logging.getLogger(__name__)
+
+
+class PagePlanner(BaseModel):
+    """Page plan model."""
+
+    title: str = Field(
+        ...,
+        description="""
+        Title of the page. If no description is provided,
+        make a concise and descriptive title from the components.
+        """,
+    )
+    page_id: str = Field(..., description="Unique identifier for the page being planned.")
+    components_plan: List[ComponentPlan] = Field(
+        ..., description="List of components. Must contain at least one component."
+    )
+    controls_plan: List[ControlPlan] = Field([], description="Controls of the page.")
+    layout_plan: LayoutPlan = Field(None, description="Layout of components on the page.")
+    unsupported_specs: List[str] = Field(
+        [],
+        description="""
+        List of unsupported specs. If there are any unsupported specs,
+        list them here. If not, leave this as an empty list.
+        """,
+    )
+
+    _components: List[Union[vm.Card, vm.AgGrid, vm.Figure]] = PrivateAttr()
+    _controls: List[vm.Filter] = PrivateAttr()
+    _layout: vm.Layout = PrivateAttr()
+
+    @validator("components_plan")
+    def _check_components_plan(cls, v):
+        if not v:
+            raise ValueError("A page must contain at least one component.")
+        return v
+
+    @validator("unsupported_specs")
+    def _check_unsupported_specs(cls, v, values):
+        title = values.get("title", "Unknown Title")
+        if v:
+            logger.warning(f"\n ------- \n Unsupported specs on page <{title}>: \n {v}")
+            return []
+
+    @root_validator(allow_reuse=True)
+    def validate_component_id_unique(cls, values):
+        """Validate the component id is unique."""
+        components = values.get("components_plan", [])
+        component_ids = [comp.component_id for comp in components]
+        duplicates = [id for id, count in Counter(component_ids).items() if count > 1]
+        if duplicates:
+            raise ValidationError(f"Component ids must be unique. Duplicated component ids: {duplicates}")
+        return values
+
+    def __init__(self, **data):
```

Comment:
Why do we need this custom `__init__`?

Reply:
to make those 3 private attributes visible, otherwise got attribute error

Reply:
Hm let's revisit this